source/agora/network/Crawler.d

```diff
@@ -0,0 +1,570 @@
+        ref NodeInfo node_info, bool is_ipv4) @safe nothrow
+    {
+        CrawlResult crawl_result;
+        foreach (i, ref field; crawl_result.tupleof)
+            static if (i == 6)
+                break;
+            else static if (is(typeof(field) == string))
+                field = extracted_values[i];
+            else static if (is(typeof(field) == TimePoint))
+                field = this.clock.networkTime();
+            else static assert(0);
+
+        crawl_result.os = node_info.os;
+        crawl_result.client_name = node_info.client_name;
+        crawl_result.client_ver = node_info.client_ver;
+        crawl_result.height = node_info.height;
+        crawl_result.is_ipv4 = is_ipv4;
+
+        return crawl_result;
+    }
+
+    unittest
+    {
+        import std.algorithm : map;
+        import std.array : array;
+        import std.conv : to;
+
+        auto mock_clock = new MockClock(5);
+        auto crawler = new Crawler(null, mock_clock, Config(), null);
+        NodeInfo node_info;
+        node_info.os = "6";
+        node_info.client_name = "7";
+        node_info.client_ver = "8";
+        node_info.height = Height(9);
+
+        CrawlResult expected;
+        expected.continent = "0";
+        expected.country = "1";
+        expected.city = "2";
+        expected.latitude = "3";
+        expected.longitude = "4";
+        expected.crawl_time = 5;
+        expected.os = "6";
+        expected.client_name = "7";
+        expected.client_ver = "8";
+        expected.height = Height(9);
+        expected.is_ipv4 = true;
+
+        assert(crawler.populateCrawlResult(iota(5).map!(i => i.to!string()).array(), node_info, true) == expected);
+    }
+
+/***************************************************************************
+
+        Gets the information about the network and it's reachable nodes.
+
+        The information includes the network address and the geograpical location
+        of the nodes including the continent/country/city/latitude/longitude.
+
+        API:
+            GET /network_info
+
+        Returns:
+            information about the network and about it's nodes
+
+
+    ***************************************************************************/
+
+
+    public CrawlResultHolder getNetworkInfo () @safe @nogc nothrow pure
+    {
+        return crawl_result_holder;
+    }
+
+    /***************************************************************************
+
+        Creates and returns a newly created NodeLocator object.
+
+        Returns:
+            a newly created NodeLocator object
+
+    ***************************************************************************/
+
+    protected INodeLocator makeNodeLocator ()
+    {
+        return new NodeLocatorGeoIP(this.config.node.ipdb_path);
+    }
+
+    // Start the Crawler
+    public void start ()
+    {
+        import std.array : array;
+
+        if (!this.config.node.collect_network_statistics)
+            return;
```

Comment:
I think it is cleaner to try to push the details down to lower level classes.

I would rather see:
```D
crawler.start()
```
in FullNode.d
instead of
```D
if (!this.config.node.collect_network_statistics)
{
	crawler = new Crawler(....);
	crawler.start()
}
```
and then checking
```D
if (crawler != null)
```D
everywhere in the FullNode if we want to use the Crawler

Reply:
> I think it is cleaner to try to push the details down to lower level classes.

As a general principle, I agree. But in this case, the crawler is something that we "fire and forget", as the node does not request it once it's instantiated. The only moment we need to coordinate with the crawler is in `start` and `stop`, and so two `if`s aren't really a concern.

Reply:
> > I think it is cleaner to try to push the details down to lower level classes.
> 
> As a general principle, I agree. But in this case, the crawler is something that we "fire and forget", as the node does not request it once it's instantiated. The only moment we need to coordinate with the crawler is in `start` and `stop`, and so two `if`s aren't really a concern.

this is in FullNode:

```D
    /// GET: /network_info
    public override CrawlResultHolder getNetworkInfo () @safe
    {
        this.endpoint_request_stats
            .increaseMetricBy!"agora_endpoint_calls_total"(1, "network_info", "http");
        return this.crawler.getNetworkInfo();
    }
```

now we will need to do a check if there is indeed a crawler instance. This check didn't even needed to be done in in crawler.getNetworkInfo before as we just returned an empty struct/

anyway, I can make the change